Signal/test/textsecure/VersionMigrationsTests.m

```diff
@@ -0,0 +1,163 @@
++(void) migrateKeyingStorageToVersion2Dot0;
+@end
+
+
+@interface SignalKeyingStorage(Testing)
++(void)storeData:(NSData*)data forKey:(NSString*)key;
++(NSData*)dataForKey:(NSString*)key;
++(void)storeString:(NSString*)string forKey:(NSString*)key;
++(NSString*)stringForKey:(NSString*)key;
+@end
+
+@interface TSDatabaseView(Testing)
++ (BOOL)threadShouldBeInInbox:(TSThread*)thread;
+@end
+
+
+@interface VersionMigrationsTests : XCTestCase
+@property (nonatomic,strong) NSString* localNumber;
+@property (nonatomic,strong) NSString* passwordCounter;
+@property (nonatomic,strong) NSString* savedPassword;
+
+@property (nonatomic,strong) NSData* signalingMacKey;
+@property (nonatomic,strong) NSData* signalingCipherKey;
+@property (nonatomic,strong) NSData* zidKey;
+@property (nonatomic,strong) NSData* signalingExtraKey;
+@property (nonatomic,strong) NSMutableArray* recentCalls;
+@end
+
+@implementation VersionMigrationsTests
+
+- (void)setUp {
+    [super setUp];
+    [Environment setCurrent:[Release unitTestEnvironment:@[]]];
+
+
+    _localNumber = @"+123456789";
+    _passwordCounter = @"20";
+    _savedPassword = @"muchlettersverysecure";
+    _signalingMacKey = [Cryptography generateRandomBytes:8];
+    _signalingCipherKey = [Cryptography generateRandomBytes:8];
+    _zidKey = [Cryptography generateRandomBytes:8];
+    _signalingExtraKey = [Cryptography generateRandomBytes:8];
+
+    // setup the keys
+    [UICKeyChainStore setString:_localNumber forKey:LOCAL_NUMBER_KEY];
+    [UICKeyChainStore setString:_passwordCounter forKey:PASSWORD_COUNTER_KEY];
+    [UICKeyChainStore setString:_savedPassword forKey:SAVED_PASSWORD_KEY];
+    [UICKeyChainStore setData:_signalingMacKey forKey:SIGNALING_MAC_KEY];
+    [UICKeyChainStore setData:_signalingCipherKey forKey:SIGNALING_CIPHER_KEY];
+    [UICKeyChainStore setData:_zidKey forKey:ZID_KEY];
+    [UICKeyChainStore setData:_signalingExtraKey forKey:SIGNALING_EXTRA_KEY];
+
+    // setup the recent calls
+    RecentCall* r1 = [RecentCall recentCallWithContactID:123
+                                              andNumber:testPhoneNumber1
+                                            andCallType:RPRecentCallTypeIncoming];
+    RecentCall* r2 = [RecentCall recentCallWithContactID:456
+                                              andNumber:testPhoneNumber2
+                                            andCallType:RPRecentCallTypeMissed];
+    
+    r2.isArchived = YES;
+    
+    _recentCalls = [[NSMutableArray alloc] initWithObjects:r1,r2,nil];
+
+    NSUserDefaults *defaults = NSUserDefaults.standardUserDefaults;
+    NSData *saveData = [NSKeyedArchiver archivedDataWithRootObject:_recentCalls.copy];
+    [defaults setObject:saveData forKey:RECENT_CALLS_DEFAULT_KEY];
+    [defaults synchronize];
+
+}
+
+- (void)tearDown {
+    // Put teardown code here. This method is called after the invocation of each test method in the class.
+    [super tearDown];
+}
+
+- (void)testMigrateKeyingStorageToVersion2Dot0 {
```

Comment:
this is the most important tests, which tests that the key material was migrated correctly and deleted correctly. passes